elephant/joint_decorrelation/joint_decorrelation.py

```diff
@@ -0,0 +1,139 @@
+
+    def _fit(self, c0, c1, verbose):
+
+        todss, fromdss, pwr0, pwr1 = dss0(c0, c1, keep1=self.n_max_components,
+                                          keep2=self.explained_variance_threshold)
+
+        self.todss = todss
+        self.n_components = self.todss.shape[1]
+        self.fromdss = fromdss
+        self.pwr0 = pwr0
+        self.pwr1 = pwr1
+        self.power_ratio = self.pwr1/self.pwr0
+
+        return self
+
+    def transform(self, data):
+
+        if data.shape[0] != self.todss.shape[0]:
+            raise ValueError('The first dimension of `data` should be equal to the number of channels.')
+
+        return np.dot(data.T, self.todss).T
+
+    def fit_transform(self, baseline_data=None, bias_filtered_data=None,
+            baseline_covmat=None, bias_filtered_covmat=None, verbose=False):
+        self.fit(baseline_data=baseline_data,
+                 bias_filtered_data=bias_filtered_data,
+                 baseline_covmat=baseline_covmat,
+                 bias_filtered_covmat=bias_filtered_covmat,
+                 verbose=verbose)
+        return self.transform(baseline_data)
+
+    def project_out(self, data, components_to_discard=[],
+                    power_ratio_threshold=None):
+        """
+        project_out here means to project from dss space to original space leaving specific components out of the data
+
+        components_to_discard should be a list of indices of component not to use for the projection
+        """
+
+        if data.shape[0] != self.todss.shape[0]:
+            raise ValueError('The first dimension of `data` should be equal to the number of channels.')
+        if not isinstance(components_to_discard, list):
+            raise ValueError('`components_to_discard` should be a list.')
+        if len(components_to_discard) > self.n_components:
+            raise ValueError('`components_to_discard` should be empty or a subset of all components.')
+        if len(components_to_discard) != 0 and max(components_to_discard) >= self.n_components:
+            raise ValueError('The indices in `components_to_discard` cannot exceed {} (`n_components`)'.format(self.n_components))
+
+        if len(components_to_discard) != 0 and power_ratio_threshold != None:
+            raise ValueError('Either one of `components_to_discard` or `power_ratio_threshold` has to be specified.')
+
+        if len(components_to_discard) == 0:
+            components_to_discard = np.where(self.power_ratio > power_ratio_threshold)[0]
+
+        mask = np.ones(self.n_components, dtype=bool)
+        mask[components_to_discard] = False
+        return self._project(data, mask)
+
+    def _project(self, data, mask):
+        if np.all(mask) == True:
+            return data
+        e_diag = np.zeros(self.n_components)
+        e_diag[mask] = 1
+        e = np.diag(e_diag)
+        return np.dot(np.dot(np.dot(data.T, self.todss), e), self.fromdss).T
+
+    def project_in(self, data, components_to_keep=[],
```

Comment:
this function is very similar to `project_out`
don't duplicate the code
think how to reuse the code
merge the functionality, if needed